packages/core/src/utilities/VoxelManager.ts

```diff
@@ -302,6 +303,74 @@ export default class VoxelManager<T> {
     voxelManager.sourceVoxelManager = sourceVoxelManager;
     return voxelManager;
   }
+
+  /**
+   * Creates a history remembering voxel manager.
+   * This will remember the original values in the voxels, and will apply the
+   * update to the underlying source voxel manager.
+   */
+  public static createLazyVoxelManager<T>(
+    dimensions: Point3,
+    layerFactory: (width: number, height: number) => T
+  ): VoxelManager<T> {
+    const map = new Map<number, T>();
+    const [width, height, depth] = dimensions;
+    const layerSize = width * height;
+
+    const voxelManager = new VoxelManager(
+      dimensions,
+      (index) => map.get(Math.floor(index / layerSize))?.[index % layerSize],
+      function (index, v) {
+        const k = Math.floor(index / layerSize);
+        let layer = map.get(k);
+        if (!layer) {
+          layer = layerFactory(width, height);
+          map.set(k, layer);
+        }
+        layer[index % layerSize] = v;
+      }
+    );
+    voxelManager.map = map;
+    return voxelManager;
+  }
+
+  /**
+   * Creates a RLE based voxel manager.  This is effective for storing
+   * segmentation maps.
+   */
+  public static createRLEVoxelManager<T>(dimensions: Point3): VoxelManager<T> {
+    const [width, height, depth] = dimensions;
+    const map = new RLEVoxelMap<T>(width, height, depth);
+
+    const voxelManager = new VoxelManager(
+      dimensions,
+      (index) => map.get(index),
+      (index, v) => map.set(index, v)
+    );
+    voxelManager.map = map;
+    return voxelManager;
+  }
+
+  /**
+   * This method adds a voxelManager instance to the image or volume object
+   * where the object added is of type:
+   * 1. RLE map if the scalar data is missing
+   * 2. Volume map for the imageId scalarData
+   */
+  public static addInstanceToImage(image: IImage) {
+    const { width, height } = image;
+    const scalarData = image.getPixelData();
+    if (scalarData.length >= width * height) {
+      image.voxelManager = VoxelManager.createVolumeVoxelManager(
+        [width, height, 1],
+        scalarData
+      );
+      return;
+    }
+    image.voxelManager = VoxelManager.createRLEVoxelManager([width, height, 1]);
+    // Some return value which can be used as the cache size.
+    return 1024;
```

Comment:
addInstanceToImage as a static method returning 1024 is odd

Reply:
It isn't a good idea, but the cache doesn't really deal with values which change sizes.  Let me know if you have a better idea about how to manage the size here.